plugins/guisupport/guisupport.cpp

```diff
@@ -499,29 +501,33 @@ QIcon GuiSupport::createIcon(const QIcon &oldIcon, QWindow *w)
         gammarayIcon.addFile(QLatin1String(":/gammaray/images/gammaray-inject-128.png"));
     }
 
-    const QIcon &orgIcon = m_originalIcons[w];
-    if (!oldIcon.isNull() && (!orgIcon.isNull() || orgIcon.cacheKey() == oldIcon.cacheKey())) {
+    // Make sure to override the window icon with the app icon if it was initialized with it
+    if (w) {
+        const auto it = m_originalIcons.constFind(qApp);
+        const QIcon appIcon = it != m_originalIcons.constEnd() ? it.value() : qApp->windowIcon();
+        if (oldIcon.cacheKey() == appIcon.cacheKey() || oldIcon.cacheKey() == qApp->windowIcon().cacheKey()) {
```

Comment:
Why do you return when the oldIcon is the same as the app icon?
Wouldn't that be exactly the case where we want to change the icon?

Reply:
For window icons implicitly set from QApplication::windowIcon() there is Qt implementation issues in QWindow, see QTBUG-63340.
Therefore we would not want to over-overlay implicit icons which are already marked by the overlay.